src/InfluxDB2/Client.php

```diff
@@ -42,6 +47,16 @@ public function __construct(array $options)
         $this->options = $options;
     }
 
+    private function getGuzzleApi(): GuzzleApi
+    {
+        return $this->guzzle_api ?? new GuzzleApi($this->options);
+    }
+
+    private function getCurlApi(): CurlApi
+    {
+        return $this->curl_api ?? new CurlApi($this->options);
+    }
+
```

Comment:
Please, init the `default_api` as part of constructor:

```suggestion
        $this->options = $options;
        if (array_key_exists("default_api", $options)) {
            $this->default_api = $this->options["default_api"];
        } else {
            $this->default_api = new GuzzleApi($this->options);
        }
    }
```

Then an user can init the client with their own implementation:

```php
client = new Client([
              "url" => "http://localhost:8086",
              "token" => "my-token",
              "bucket" => "my-bucket",
              "default_api" => new MyDefaultApi()
         ]);
```

Reply:
From the perspective of a dependency injection container, I can imagine "hiding" objects inside array parameters can cause difficulties. Perhaps setter injection would provide a better interface here? So something like

```php
public function setDefaultApi(DefaultApi $default_api)
{
    $this->default_api = $default_api;
}

private function getDefaultApi(): DefaultApi
{
    return $this->default_api ?? new GuzzleApi($this->options);
}
```

This essentially provides the same experience as you are suggesting, in that Guzzle is the default for anyone who doesn't care, and if you want to you can replace it with something else.

Reply:
Yeah, using setter injection is also possibility how to do it, but the `getDefaultApi` should avoid creating new `GuzzleApi` in every call.

Reply:
I now see that that's not what my suggested code is doing, but that was indeed the intention :)